IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m

```diff
@@ -444,16 +438,60 @@ - (BOOL)saveAppMetadata:(nonnull MSIDAppMetadataCacheItem *)metadata
+                                                       context:(nullable id<MSIDRequestContext>)context
+                                                         error:(NSError * _Nullable * _Nullable)error
 {
-    assert(key);
+    assert(query);
+
+    MSID_LOG_VERBOSE(context, @"(Default cache) Get app metadata for query %@", query.logDescription);
+
+    if (query.exactMatch)
+    {
+        return [_dataSource appMetadataWithKey:query serializer:_serializer context:context error:error];
+    }
+
+    NSArray<MSIDAppMetadataCacheItem *> *appMetadataEntries = [_dataSource appMetadataEntriesWithKey:query
+                                                                                          serializer:_serializer
+                                                                                             context:context
+                                                                                               error:error];
+
+    for (MSIDAppMetadataCacheItem *appMetadata in appMetadataEntries)
+    {
+        if ([appMetadata matchesWithClientId:query.clientId
+                                 environment:query.environment
+                          environmentAliases:query.environmentAliases])
+        {
+            return appMetadata;
+        }
+    }
+
+    return nil;
+}
+
+- (BOOL)removeAppMetadata:(nonnull MSIDAppMetadataCacheItem *)appMetadata
+                  context:(nullable id<MSIDRequestContext>)context
+                    error:(NSError * _Nullable * _Nullable)error
+{
+    assert(appMetadata);
+    
+    MSID_LOG_VERBOSE(context, @"(Default cache) Removing app metadata with clientId %@, environment %@", appMetadata.clientId, appMetadata.environment);
     
-    MSID_LOG_VERBOSE(context, @"(Default cache) Get app metadata for key %@", key.logDescription);
-    MSID_LOG_VERBOSE_PII(context, @"(Default cache) Get app metadata for key %@", key.piiLogDescription);
+    MSIDAppMetadataCacheKey *key = [[MSIDAppMetadataCacheKey alloc] initWithClientId:appMetadata.clientId
+                                                                         environment:appMetadata.environment
+                                                                            familyId:appMetadata.familyId
+                                                                         generalType:MSIDAppMetadataType];
+    
+    return [_dataSource removeItemsWithKey:key context:context error:error];
+}
+
+- (nullable NSArray<MSIDAppMetadataCacheItem *> *)getAppMetadataEntries:(nonnull MSIDAppMetadataCacheQuery *)query
+                                                                context:(nullable id<MSIDRequestContext>)context
+                                                                  error:(NSError * _Nullable * _Nullable)error
+{
+    MSID_LOG_VERBOSE(context, @"(Default cache) Get all app metadata entries with type %@", [MSIDGeneralCacheItemTypeHelpers generalTypeAsString:query.generalType]);
```

Comment:
Please log additional things from query match parameter, and do additional matching if necessary. See my comment above.